Add guide on debugging providers. (#1584)
hashicorp/terraform-website: content/source/docs/extend/debugging.html.md

---
layout: "extend"
page_title: "Extending Terraform: Debugging Providers"

> sidebar_current: "docs-extend-debugging-providers"

Reviewer: Can totally omit `sidebar_current` if you want. It no longer does anything.

description: |-
  How to debug providers.
---

# Debugging Providers

> 
> Terraform's providers are complicated programs in their own right, and
> sometimes they don't end up working the way you expect them to. In those
> situations, it's not always obvious how to get information about what's
> actually happening inside a provider or why things are behaving the way they
> do.

Reviewer 2: ```suggestion
```

I don't think we need this motivating paragraph. If you've reached this document then you probably know what debugging is, and the next paragraph will tell you whether this document is meant for you.

it.

There are two available approaches to debugging Terraform providers. We'll talk
about each of them separately and in turn.

## Log-Based Debugging

Log-based debugging is a method of using logging calls to record what is
happening in a provider as it happens, and then examining that record to piece
together what happened after the fact. Usefully, end users can capture logs and
share them with developers, allowing developers to sometimes debug providers
without needing access to the environment or configuration. Unfortunately,
developers need to think ahead about what information they'll need when
debugging, because the log lines need to be injected into the provider when
it's built and cannot be inserted into the binary after it has been built. This
makes it harder to ask new questions when doing log-based debugging, as the log
line needs to be added and the provider needs to be recompiled before new
information can be surfaced.

### Inserting Log Lines Into A Provider

Terraform providers all log to `stderr`. Any output to `stderr` will be passed
over the gRPC protocol to Terraform, which will then write it to its
destination.

> ~> **NOTE:** Don't write to `stdout` on any Terraform providers! It can cause

Reviewer: Yellow boxes -> "important" instead of "note".

```suggestion
~> **Important:** Don't write to `stdout` on any Terraform providers! It can cause
```

   problems with our plugin system, which uses `stdout` as a communication
   channel.

> It is recommended to use the standard library `log` package to achieve this,
> using `log.Println` or `log.Printf` or similar functions.

Reviewer: ```suggestion
We recommend using the standard library `log` package to achieve this,
via `log.Println` or `log.Printf` or similar functions.
```

Log lines must be prefixed with the level of the log in square brackets. Valid

> levels are:

Reviewer: ```suggestion
levels, in decreasing order of severity, are:
```

Experienced programmers probably already guessed that, but might as well take the chance to teach when it arises.

* ERROR
* WARN
* INFO
* DEBUG
* TRACE

A full log line, then, may look like this:

```go
log.Println("[DEBUG] Something happened!")
```

### Turning On Logging

Terraform controls which log levels are recorded for all providers and
Terraform itself using the `TF_LOG` environment variable. Set it to `TRACE` to
output all log lines, `DEBUG` to output all log lines except `TRACE` lines,
etc.

~> **NOTE:** prior to v0.15.0 of Terraform, levels besides `TRACE` may not be
   entirely reliable. See [this
   PR](https://github.com/hashicorp/terraform/pull/26632) for more details.

By default, log output will be recorded to `stderr` when `TF_LOG` is set. To
write log output to a file instead, set `TF_LOG_PATH` to a file's path. Log
output will be appended there. It's worth noting that this file will not be
truncated before log output is written.

#### Separating Out Provider Logs

Terraform can have some verbose logging and it can be difficult to parse log

> files when both Terraform and providers are writing to the same log file. To
> address this, following the 0.15.0 release of Terraform, the `TF_LOG_CORE`
> environment variable and `TF_LOG_PROVIDER` environment variable can now be set
> to different levels. Set either to `off` to turn off its logging entirely.

Reviewer: ```suggestion
files when both Terraform and providers are writing to the same log file. In
Terraform 0.15.0 and later, you can address this by setting the `TF_LOG_CORE`
and `TF_LOG_PROVIDER` environment variables to different log levels.
Set either to `off` to turn off its logging entirely.
```

- Reducing passive voice.
- Putting version requirement earlier in that sentence.

## Debugger-Based Debugging

> Debugger-based debugging is a method of using debugging tools like
> [delve](https://github.com/go-delve/delve) to inspect what is happening in a
> provider as it is happening, often using breakpoints to pause execution of the
> provider and examine the values of variables. This method of debugging must be
> done contemporaneously; the developer doing the debugging needs to actively run
> Terraform using the appropriate configuration and in the appropriate
> environment to induce the behavior being examined. It is therefore most useful
> when a bug has been reliably reproduced. It does, however, enable developers to
> ask arbitrary questions and step through provider executions, exploring what is
> happening in the provider as part of the execution.

Reviewer 3: ```suggestion
Debugger-based debugging is a method of using a tool like
[delve](https://github.com/go-delve/delve) to inspect what is happening in a
provider as it is happening, often using breakpoints to pause execution of the
provider and examine the values of variables. This method of debugging must be
done contemporaneously; the developer doing the debugging needs to actively run
Terraform using the appropriate configuration and in the appropriate
environment to induce the behavior being examined. It is therefore most useful
when a bug is reliably reproducible. This level of analysis enables developers to
ask arbitrary questions and step through provider executions, allowing them to 
explore what is happening in the provider during runtime.
```

> ~> **NOTE**: Debugger-based debugging only works with Terraform versions

Reviewer: ```suggestion
-> **Note**: Debugger-based debugging only works with Terraform versions
```

This one can be a blue box, since it's not warning you of any consequences.

   0.12.26 and higher.

### Enabling Debugging In A Provider

> Before a debugger can attach to a provider, the provider must be using version
> 2.0.0 of the Terraform Plugin SDK. The plugin must also be started in debug

Reviewer 2: ```suggestion
Debugging is available for providers using Terraform Plugin SDK versions 2.0.0
and above. The plugin must also be started in debug
```

> mode, called `plugin.Debug` instead of `plugin.Serve`. It is recommended you

Reviewer: ```suggestion
mode, called `plugin.Debug` instead of `plugin.Serve`. We recommend that you
```

passive voice

enable this using a flag, as the provider should use `plugin.Serve` under
normal operation, when not being debugged.

```go
func main() {
	var debugMode bool

	flag.BoolVar(&debugMode, "debug", false, "set to true to run the provider with support for debuggers like delve")
	flag.Parse()

	opts := &plugin.ServeOpts{ProviderFunc: provider.New}

	if debugMode {
		// TODO: update this string with the full name of your provider as used in your configs
		err := plugin.Debug(context.Background(), "registry.terraform.io/my-org/my-provider", opts)
		if err != nil {
			log.Fatal(err.Error())
		}
		return
	}

	plugin.Serve(opts)
}
```

> It is important that a provider only be started in debug mode when you intend

Reviewer 2: ```suggestion
It is important to start a provider in debug mode only when you intend
```

Alternative to @nfagerlund's suggestion for those with a split-infinitive intolerance.

> It is important that a provider only be started in debug mode when you intend
> to debug it, as its behavior will change in minor ways from normal operation of
> providers. The main differences are:

Reviewer: ```suggestion
It is important to only start a provider in debug mode when you intend
to debug it, as its behavior will change in ways that would be undesirable in 
normal use. The main differences are:
```
